### DISCUSSION TITLE
Using with an external vault Provider like AWS secrets manager or Hashicorp Vault

### DISCUSSION THREAD

**Discussion by louim** [palkan/anyway_config#150] [Q&A]
Hey! I was looking into integrating with a vault provider like AWS Secrets Manager or Hashicorp Vault. Those services usually require that you provide the names of the individual secrets you want to load. There's usually no endpoint to load all secrets at once. Is this concept fundamentally opposed to how AnywayConfig work? If thats not the case, how would one provide the loader with the secrets that are required to be fetched?

**Comment by palkan** [comment]
Hey,

> Is this concept fundamentally opposed to how AnywayConfig work?

It’s not something we had in mind but we should be able to figure out the workaround. 

We can add support for accessing the list of configuration parameters to loaders. Then, the corresponding loader will have to iterate and try to load every key. Doesn’t sound pretty good though.

Can we somehow specify keys to load? 
Well, that would contradict the idea of separating configuration model (class) from data sources. And that wouldn’t help with providing values for third-party configs.

One solutions that comes into my mind is as follows:

- Keep a separate file (YAML) containing the list of keys to be fetched from an external vault for every config (application and third-party).

- Write a loader that uses this configuration to decide which keys to fetch for a given config class.

This way, config classes stay unaware of any providers.



----

Related Vault issue:
https://github.com/hashicorp/vault/issues/272

**Reply by louim** [reply]
Thanks for the reply! My thinking was roughly like so:
- Configuration parameters are passed to the loader
- The loader fetch the secrets list from the external provider. As an example, using [list secrets](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SecretsManager/Client.html#list_secrets-instance_method) for AWS.
- The loader fetch the actual secrets against the list of secrets contained in the vault and the configuration parameters passed. For loaders that implement batch fetch, like [AWS secrets manager](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SecretsManager/Client.html#batch_get_secret_value-instance_method), it could be loaded in batch. Otherwise the secrets are fetched individually in parallel.

This would remove the need to maintain a separate list of secrets to be fetched for the external provider. Such configuration could be supported for cases where performance requires it, but I think for most cases it wouldn't be required. What do you think?

So the first step to get this working would be:

> We can add support for accessing the list of configuration parameters to loaders.

I'd be happy to contribute this with some guidance if it's not too involved.

**Reply by palkan** [reply]
> The loader fetch the secrets list from the external provider

Oh, that's great. Didn't know about that. I think, in this case no changes to the core needed, just a custom loader and some naming convention for secrets (so the loader could pick up only the settings relevant to the current config).

**Comment by timwis** [comment]
Say, what about using yaml tags? E.g. where your secret is stored in hashicorp vault:

```yml
# example.yml

production:
  endpoint: https://api.example.com
  api_key: !vault example-api-key
```
